Upstream: http://lists.gnu.org/archive/html/grub-devel/2009-08/msg00694.html
 (in part; see thread)
Description: Allow Shift to interrupt 'sleep --interruptible'.
UbuntuSpecific: Upstream would like to consider this at more length. See
 http://lists.gnu.org/archive/html/grub-devel/2009-08/msg00718.html, and the
 rest of the thread for context.

diff -ur grub2-1.98-ylmf2/commands/sleep.c grub2-1.98-ylmf2.new/commands/sleep.c
--- grub2-1.98-ylmf2/commands/sleep.c	2010-03-07 04:51:37.000000000 +0800
+++ grub2-1.98-ylmf2.new/commands/sleep.c	2011-05-25 16:29:04.019542937 +0800
@@ -44,6 +44,20 @@
   grub_printf ("%d    ", n);
 }
 
+static int
+grub_check_keyboard (void)
+{
+  int mods = grub_getkeystatus ();
+  if (mods >= 0 && (mods & GRUB_TERM_STATUS_SHIFT) != 0)
+    return 1;
+
+  if (grub_checkkey () >= 0 &&
+      GRUB_TERM_ASCII_CHAR (grub_getkey ()) == GRUB_TERM_ESC)
+    return 1;
+
+  return 0;
+}
+
 /* Based on grub_millisleep() from kern/generic/millisleep.c.  */
 static int
 grub_interruptible_millisleep (grub_uint32_t ms)
@@ -53,8 +67,7 @@
   start = grub_get_time_ms ();
 
   while (grub_get_time_ms () - start < ms)
-    if (grub_checkkey () >= 0 &&
-	GRUB_TERM_ASCII_CHAR (grub_getkey ()) == GRUB_TERM_ESC)
+    if (grub_check_keyboard ())
       return 1;
 
   return 0;
